### PR TITLE
python: Extend Notification class

### PR DESCRIPTION
**Pull request by vmytnykx**
Extended python Notification class to support metadata in the same
way as it is implemented in Values class. The Values metadata class
implementation has been used as a example/source.

### CODE REVIEW

**Comment by maryamtahhan** [comment]
Hi Folks
Any updates on this PR?

BR
Maryam

**Comment by maryamtahhan** [comment]
@octo @rubenk is there anything we can do here to expedite this review?

Thanks in advance
Maryam

**Comment by rubenk** [comment]
cc @trenkel

File: src/collectd-python.pod [hunk 1]
```diff
@@ -475,7 +475,7 @@ Methods defined here:
 
 =over 4
 
-=item B<dispatch>([type][, values][, plugin_instance][, type_instance][, plugin][, host][, time][, interval]) -> None.  Dispatch a value list.
+=item B<dispatch>([type][, message][, plugin_instance][, type_instance][, plugin][, host][, time][, severity][, meta]) -> None.  Dispatch a notification.
```
Comment (octo) [inline src/collectd-python.pod]: Thanks for noticing this! I've cherry-picked this fix into *collectd-5.7* as be906f960ede42f72317eab6aa6a5f116857a484.

File: src/python.c [hunk 2]
```diff
@@ -510,6 +510,35 @@ static int cpy_notification_callback(const notification_t *notification,
...
+  for (notification_meta_t *meta = notification->meta;
+       meta != NULL; meta = meta->next) {
+    PyObject *temp = NULL;
+    if (meta->type == NM_TYPE_STRING) {
+      temp = cpy_string_to_unicode_or_bytes(meta->nm_value.nm_string); /* New reference. */
+      PyDict_SetItemString(dict, meta->name, temp);
+      Py_XDECREF(temp);
+    } else if (meta->type == NM_TYPE_SIGNED_INT) {
+      PyObject *sival = PyLong_FromLongLong(meta->nm_value.nm_signed_int);
+      temp = PyObject_CallFunctionObjArgs((void *)&SignedType, sival,
```
Comment (octo) [inline src/python.c]: Why can't you just do this:?

```c
PyDict_SetItemString(dict, meta->name,
                     PyLong_FromLongLong(meta->nm_value.nm_signed_int));
```

Dito for `NM_TYPE_UNSIGNED_INT` below.

File: src/python.c [hunk 3]
```diff
@@ -510,6 +510,35 @@ static int cpy_notification_callback(const notification_t *notification,
...
+    } else if (meta->type == NM_TYPE_UNSIGNED_INT) {
+      PyObject *uval = PyLong_FromUnsignedLongLong(meta->nm_value.nm_unsigned_int);
+      temp = PyObject_CallFunctionObjArgs((void *)&UnsignedType, uval,
+                                          (void *)0); /* New reference. */
+      PyDict_SetItemString(dict, meta->name, temp);
+      Py_XDECREF(temp);
+    } else if (meta->type == NM_TYPE_DOUBLE) {
+      temp = PyFloat_FromDouble(meta->nm_value.nm_double); /* New reference. */
+      PyDict_SetItemString(dict, meta->name, temp);
+      Py_XDECREF(temp);
```
Comment (octo) [inline src/python.c]: Why do you call `Py_XDECREF()` on the return value of `PyFloat_FromDouble()` but not `PyLong_From[Unsigned]LongLong()`? According to the docs, all of these return a "new reference", don't they need to be treated equally?

File: src/pyvalues.c [hunk 4]
```diff
@@ -42,6 +50,26 @@
     PyMem_Free(host);                                                          \
   } while (0)
 
+#define CPY_BUILD_META_FUNC(meta_type, func, val_type)                         \
```
Comment (octo) [inline src/pyvalues.c]: This is some serious macro magic. Please move the macro definition and invocation closer together so it's easier to understand what's going on.

File: src/pyvalues.c [hunk 5]
```diff
@@ -541,9 +568,22 @@ static meta_data_t *cpy_build_meta(PyObject *meta) {
     Py_DECREF(key);
   }
   Py_XDECREF(l);
+  return 0;
+}
+
+CPY_BUILD_META_HANDLER(meta_data, meta_data_t);
+CPY_BUILD_META_HANDLER(plugin_notification_meta, notification_t);
+
+static meta_data_t *cpy_build_value_meta(PyObject *meta) {
```
Comment (octo) [inline src/pyvalues.c]: I think this pretty wrapper should get the pretty `cpy_build_meta()` name. The other function could be renamed to `cpy_build_meta_generic()` to underline that it works with more than one meta type.

File: src/python.c (hunk 2 above)
Comment (vmytnykx) [reply src/python.c]: It's possible but collectd python plugin doesn't use `long` type directly. It uses collectd python types (`Signed` and `Unsigned`) instead. Thus, we have to convert python long/unsigned long type into `Signed`/`Unsigned` types. Same has been done for python value.
https://github.com/collectd/collectd/blob/master/src/collectd-python.pod

Comment (octo) [reply src/python.c]: That makes sense, thank you! I didn't realize right away that the `Signed` / `Unsigned` types are *not* built-in literals but classes.

File: src/python.c (hunk 3 above)
Comment (vmytnykx) [reply src/python.c]: This is a good question! Based o my knowledge and testing it looks like a potential memory leak. Same logic is used when metadata is created for a value, so the problem needs to be fixed in both places in this cases.

**Comment by vmytnykx** [comment]
Hi @octo, I've fixed all your comments. Thanks for your code review.

**Comment by maryamtahhan** [comment]
Thanks Evenryone!

**Pull request merged**